### ISSUE TITLE
Define policy for name collisions with Scala standard library

### ISSUE THREAD

**Issue by benhutchison** [non/cats#831]
A number of cats methods collide with methods defined by the standard library, eg `Functor.map`, `Foldable.foldRight`. Because the cats' methods pimped onto a class have lower priority than methods defined on the class itself, the cats methods cannot be readily invoked using infix syntax. This can lead to the same operation having different a implementation, and occasionally behaviour, for different container types:

```
@ Xor.right(1).map(_.toString)
res9: Xor[Nothing, String] = Right(1)
@ Vector(1).map(_.toString)
res10: Vector[String] = Vector("1")
@ BitSet(1).map(_.toString)
res11: SortedSet[String] = TreeSet("1")
```

It would help people writing applications on top of cats to have a clearly expressed policy around name collisions, or lack of, with the standard library. Two options are:

1. *Coexistence through compatibility*. Cats methods collide with the standard library, but so long as they do the same thing (almost all the time) its not a problem in practice. So any cats method sharing a method with scala collections API must have same behavior.

2. *Non-collision*. Cats methods never use the same name as the existing collections API. This has the benefit that it's obvious to library users whether they're using cats or std lib methods. It has the downside of requiring a big, disruptive change in the current cats api, and also means that many commonly used, intuitive names are off-limits, since they're taken by the std lib.

There may be other options? Whatever course is chosen, would be helpful to define the policy both for people (a) building on top of the library, and (b) adding methods to the api.

**Comment by stew** [comment]
So I could get behind this, with one caveat. For map / flatMap, I think we have to always have to provide the standard map/flatMap names so that having a Monad in scope allows you to use a for-comprehension. so for this one perhaps we can provide fmap / bind alternatives so that people can both use for comprehensions for types that otherwise wouldn't have map/flatMap, and you can have fmap/bind for when you want to use the inline syntax and make sure you are getting the cats version.

**Comment by ceedubs** [comment]
My vote would be for option 1 (which is what's currently done for the most part). Here are some reasons that come to mind:

If you are working with an abstract type such as `F[A]`, then calling `.map` is going to use `Functor[F]`. It's only when you are dealing with a concrete type that the type class won't be used, and I think that usually calling the method directly on the concrete type is fine and can potentially avoid syntax allocation. I can see an argument that it may not be obvious whether you are using a type class implementation or not, but I guess I've never had this be an issue in practice.

In places where there is a significant difference with std lib types, I do think that we should provide at least an alias that can be used to avoid collision.

As mentioned, we kind of need `map` and `flatMap` to have these names in order to be for-comprehension-friendly. I don't like the idea that if new methods are added to std-lib types in the future, that we may need to change the names of our methods again.

If people really want to make sure they are using the type class methods, then they can use [abstracted](https://github.com/travisbrown/abstracted), so they can opt-in to this sort of behavior. I suspect most people won't feel the need to do this.

Having said all of this, there's a good chance that I'm just defaulting to what I'm used to, and I probably should be more open to change :)

**Comment by stew** [comment]
@ceedubs so what about my suggestion that when we have conflicts we provide an alternative also? so we still always provide map/flatMap, but we also provide fmap/bind so that you also have a version that doesn't conflict

**Comment by stew** [comment]
also @ceedubs what about foldRight where the signature is much different that the stdlib version?

**Comment by milessabin** [comment]
I agree with @ceedubs that where we're working with an abstract type this is a non-issue. Where we're working with a concrete type I think rather than provide a differently named alias it might be better to provide a same-named method on the companion object of the type class. So instead of,

```scala
  l.catsFoldRight ...
```
we could have,
```scala
Foldable.foldRight(l) ...
```
I think this makes it very clear which operation is in play. I think the additional verbosity is fine because situations where we are using the concrete type and that type has a clashing name with the wrong semantics or signature will be relatively rare.
